### PR TITLE
Controllers doesn't take any actions when being deleted.

### PR DESCRIPTION
**Pull request by gmarek**
I started doing it for other controllers but it's not always clear to me how it should work. I'll be adding other ones as separate commits to this PR.

cc @caesarxuchao @lavalamp

### CODE REVIEW

File: pkg/controller/daemon/controller.go [hunk 1]
```diff
@@ -671,7 +671,7 @@ func (dsc *DaemonSetsController) syncDaemonSet(key string) error {
 		return err
 	}
 	dsNeedsSync := dsc.expectations.SatisfiedExpectations(dsKey)
-	if dsNeedsSync {
+	if dsNeedsSync && ds.DeletionTimestamp == nil {
```
Comment (caesarxuchao) [inline pkg/controller/daemon/controller.go]: I'm doing this in a similar way for RC, so SGTM.

Comment (caesarxuchao) [inline pkg/controller/daemon/controller.go]: For RC,  I also need to check DeletionTimestamp!=nil before adopting orphaned pods. daemon controller may need to do the same.

Comment (gmarek) [inline pkg/controller/daemon/controller.go]: The amount of boilerplate is insane...

Comment (gmarek) [inline pkg/controller/daemon/controller.go]: Am I mistaken, or adoption is not implemented for Daemons yet?

Comment (caesarxuchao) [inline pkg/controller/daemon/controller.go]: It's where you want to set the pod's Controller ref. I mean you need to stop setting controller ref if the controller is about to be deleted. In daemon set's case, we might do it in addPod().

I'll push my changes to RC later today so I can get your comments, though it's not completed :)

**Comment by gmarek** [comment]
CC @janetkuo for deployment cc @soltysh for job

**Comment by caesarxuchao** [comment]
#27600 is a WIP implementation for RC.

**Comment by soltysh** [comment]
cc @kargakis @pwittrock @mfojtik for deployments

**Comment by soltysh** [comment]
jobs LGTM

**Comment by kargakis** [comment]
Shouldn't we just not resync the object in case it is deleted? It's much simpler than handling DeletionTimestamp in the controller loop.

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -431,13 +431,14 @@ func (dc *DeploymentController) syncDeployment(key string) error {
...
 	d := obj.(*extensions.Deployment)
+
 	everything := unversioned.LabelSelector{}
 	if reflect.DeepEqual(d.Spec.Selector, &everything) {
 		dc.eventRecorder.Eventf(d, api.EventTypeWarning, "SelectingAll", "This deployment is selecting all pods. A non-empty selector is required.")
 		return nil
 	}
 
-	if d.Spec.Paused {
+	if d.Spec.Paused || d.DeletionTimestamp != nil {
```
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: A deleted deployment is different from a paused deployment; these two should be separate. I think the best fix would be to stop requeueing the deployment when it is deleted.

**Comment by gmarek** [comment]
@kargakis - I'm not certain if I understand correctly what you're saying - do you mean that we shouldn't run `syncPausedDeploymentStatus` on DeploymentController if deletionTimestamp is set? That was my original idea, but if I correctly recall my discussion with @bgrant0607 we decided that we want to prevent controllers from doing any meaningful changes while we still want to update Statuses.

**Comment by kargakis** [comment]
I mean that we shouldn't requeue at all on deletion: https://github.com/kubernetes/kubernetes/blob/d244a347474cf2b392cc8dda6746f8043077d887/pkg/controller/deployment/deployment_controller.go#L214

Deletions ought to be fast so I am not sure if there is any value in updating the status of an object but I may be wrong.

**Comment by gmarek** [comment]
@kargakis - they won't be fast soon enough, when cascading deletion will be turned on. Before deleting object X we'll wait until all dependees are deleted, which can take some time - especially that Pods do have 30s grace period by default.

**Comment by gmarek** [comment]
Thanks for the line - I'll look at that part of code.

**Comment by kargakis** [comment]
Hm, i see. Agreed

File: pkg/controller/deployment/deployment_controller.go (hunk 2 above)
Comment (kargakis) [inline pkg/controller/deployment/deployment_controller.go]: Please update this and check for the timestamp before checking for paused. https://github.com/kubernetes/kubernetes/pull/20273 changes paused deployments to do more things than simple staus updates.

**Comment by gmarek** [comment]
I didn't manage to finish this before vacation:/ Let's proceed with reviews.

**Comment by pwittrock** [comment]
Added myself as a reviewer for Deployments.

**Comment by caesarxuchao** [comment]
@janetkuo @pwittrock do you know if the deployment controller adopts existing replication controllers? Non-empty DeleteTimestamp should prevent the adoption as well.

**Comment by kargakis** [comment]
> @janetkuo @pwittrock do you know if the deployment controller adopts existing replication controllers? Non-empty DeleteTimestamp should prevent the adoption as well.

Yes, Deployments can adopt replica sets.

**Comment by gmarek** [comment]
Changed deployment logic: @pwittrock and @kargakis PTAL

**Comment by kargakis** [comment]
Deployment controller changes LGTM

**Comment by gmarek** [comment]
@mikedanese - can you take a look at DaemonController changes?

**Comment by gmarek** [comment]
Ping @mikedanese

**Comment by mikedanese** [comment]
DaemonsetController change LGTM

**Comment by gmarek** [comment]
Squashed commit. @caesarxuchao - can you apply the final LGTM?

File: pkg/controller/daemon/daemoncontroller_test.go [hunk 3]
```diff
@@ -289,6 +289,24 @@ func TestSufficentCapacityNodeDaemonLaunchesPod(t *testing.T) {
 	syncAndValidateDaemonSets(t, manager, ds, podControl, 1, 0)
 }
 
+// DaemonSets should place onto nodes with sufficient free resource
```
Comment (caesarxuchao) [inline pkg/controller/daemon/daemoncontroller_test.go]: Could you update the comment?

**Comment by caesarxuchao** [comment]
Just a nit, otherwise LGTM.

File: pkg/controller/daemon/daemoncontroller_test.go (hunk 3 above)
Comment (gmarek) [inline pkg/controller/daemon/daemoncontroller_test.go]: Yikes. Thanks for noticing.

**Comment by gmarek** [comment]
Done. @caesarxuchao PTAL

**Pull request merged**